tools/syz-declextract/syz-declextract.cpp

```diff
@@ -1,279 +1,103 @@
-    auto src = D->getSourceRange().getBegin().printToString(Context.getSourceManager());
-    if (strstr(src.c_str(), "/um/") == 0) return true;
-    for (auto file: ignore_files) {
-      if (strstr(src.c_str(), file)) return true;
+
+    auto values = llvm::dyn_cast<InitListExpr>(varDecl->getInit())->inits();
+    if (values.empty()) {
+      return;
     }
-    for (const ParmVarDecl *P : D->parameters()) {
-      auto typ = convertType(Context, P->getType());
-      if (typ == "") {
-        llvm::outs() << D->getNameInfo().getAsString() << ": UNKNOWN TYPE: " <<
-            QualType(P->getType()).getAsString() << "\n";
-        return true;
+    int argc =
+        *values[2]->getIntegerConstantExpr(*context).value().getRawData();
+    std::vector<std::string> arg_types(argc);
+    std::vector<std::string> args(argc);
+
+    if (argc) {
+      int i = 0;
+      for (const auto *item :
+           llvm::dyn_cast<InitListExpr>(
+               llvm::dyn_cast<VarDecl>(
+                   values[3]
+                       ->getAsBuiltinConstantDeclRef(*context)
+                       ->getUnderlyingDecl())
+                   ->getInit())
+               ->inits()) {
+        arg_types[i++] = std::move(*item->tryEvaluateString(*context));
       }
-    }
-    if (Generated[D->getNameInfo().getAsString()])
-      return true;
-    Generated[D->getNameInfo().getAsString()] = true;
 
-    llvm::outs() << D->getNameInfo().getAsString() << "(";
-    int i = 0;
-    for (const ParmVarDecl *P : D->parameters()) {
-      if (i)
-        llvm::outs() << ", ";
-      auto name = P->getNameAsString();
-      if (name == "") {
-        char buf[10];
-        sprintf(buf, "arg%d", i);
-        name = buf;
+      i = 0;
+      for (const auto *item :
+           llvm::dyn_cast<InitListExpr>(
+               llvm::dyn_cast<VarDecl>(
+                   values[4]
+                       ->getAsBuiltinConstantDeclRef(*context)
+                       ->getUnderlyingDecl())
+                   ->getInit())
+               ->inits()) {
+        args[i++] = std::move(*item->tryEvaluateString(*context));
       }
-      llvm::outs() << name << " " << convertType(Context, P->getType());
-      i++;
-      if (i == 9)
-        break;
     }
-    llvm::outs() << ")";
-    auto ret = convertType(Context, D->getReturnType());
-    if (ret == "HANDLE")
-      llvm::outs() << " " << ret;
-    llvm::outs() << "\n";
-    return true;
-  }
-
- private:
-  ASTContext &Context;
-  std::map<std::string, bool> Generated;
-};
-
-class DeclExtractCallConsumer : public clang::ASTConsumer {
- public:
-  explicit DeclExtractCallConsumer(ASTContext *Context)
-      : Visitor(Context) {}
 
-  virtual void HandleTranslationUnit(clang::ASTContext &Context) {
-    Visitor.TraverseDecl(Context.getTranslationUnitDecl());
+    printf("==========SYSCALL Found==========\n");
+    printf("%s\n", values[0]->tryEvaluateString(*context).value().c_str());
+    for (int i = 0; i < argc; ++i) {
```

Comment:
I think a for: loop is more preferable here.

Reply:
I created a struct to hold the parameter type and name, then used a range-based for loop.